### ISSUE TITLE
Custom branching validation error messages for a single field

### ISSUE THREAD

**Issue by bauerca** [gcanti/tcomb-form#151]
Hey y'all!

I am trying to supply distinct, custom error messages for a given form field that depend on the current value. For example, consider a string field that is required and cannot be more than 10 chars. Currently, I am using a custom error function in the tcomb Form options, but find myself repeating all the validation of the subtype inside it:

```js
var SmallStr = t.subtype(t.Str, function(s) { return s.length < 11; }, 'SmallStr');

var formOptions = {
  fields: {
    small_str: {
      error: function(value) {
        return (
          (!value && 'Must supply a string, my friend.') ||
          (value.length > 10 && 'Must be 10 chars or fewer.')
        );
      }
    }
  }
};
```

What is tcomb best practice for this sort of thing?

I would love to be able to address this at the type definition level:

```js
// Not valid tcomb syntax!!
var SmallStr = t.subtype(t.Str, {
  validate: function(value) {
    return (
      (!value && 'Must supply a string, my friend.') ||
      (value.length > 10 && 'Must be 10 chars or fewer.')
    );
  }
  displayName: 'SmallStr'
});
```

There is a stagnant issue over in tcomb that somewhat mirrors this request. https://github.com/gcanti/tcomb-validation/issues/9

Thanks everybody!

**Comment by benmonro** [comment]
+1 on this.

**Comment by gcanti** [comment]
First of all I want to apologize to @bauerca, somehow I completely missed this issue, I'm sorry.

> I would love to be able to address this at the type definition level

What about a custom combinator to keep DRY?

```js
function mycombinator(type, validate, name) {
  var Result = t.subtype(type, function (x) {
    return !t.String.is(validate(x));
  });
  Result.validate = validate;
  return Result;
}

var SmallString = mycombinator(t.String, function (value) {
  return (
    (!value && 'Must supply a string, my friend.') ||
    (value.length > 10 && 'Must be 10 chars or fewer.')
  );
}, 'SmallString');

var Foo = t.struct({
  bar: SmallString
});

var options = {
  fields: {
    bar: { error: SmallString.validate }
  }
};
```

**Comment by benmonro** [comment]
:+1:

**Comment by benmonro** [comment]
I'm trying to use tcomb isomorphically in react with fluxible.  When I call validate() on the server I want it to return the same error message as it gets on the client.  However, in order to get the error from SmallString, I have to explicitly call SmallString.getError().  Is there any way that the result of that in the ValidationResult?

i.e. using the above example:

```
var result = validate(value, Person);
```
after running that,  result.errors[0].message == 'Invalid value "xxxxxxxxx" supplied to /name: SmallString'

but I'd like it to be the result of SmallString.getError.

  Is this possible?

**Comment by gcanti** [comment]
> Is this possible?

Hi @benmonro,
It's not possible at the moment, but it's a nice feature. I opened https://github.com/gcanti/tcomb-validation/issues/23

**Comment by gcanti** [comment]
@benmonro Ok, just published tcomb-validation v2.1.1 with the new experimental feature

**Comment by bauerca** [comment]
Cool stuff! These updates look great. I'll let you close the issue/question.
